src/kv.rs

```diff
@@ -0,0 +1,241 @@
+use std::collections::HashMap;
+
+use async_nats::{jetstream::kv::Store, Client};
+use wasmbus_rpc::core::LinkDefinition;
+
+use crate::GetClaimsResponse;
+use crate::LinkDefinitionList;
+use crate::Result;
+
+pub(crate) async fn get_kv_store(
+    nc: Client,
+    lattice_prefix: &str,
+    js_domain: Option<String>,
+) -> Option<Store> {
+    let jetstream = if let Some(domain) = js_domain {
+        async_nats::jetstream::with_domain(nc, domain)
+    } else {
+        async_nats::jetstream::new(nc)
+    };
+    let bucket = format!("LATTICEDATA_{}", lattice_prefix);
+    jetstream.get_key_value(bucket).await.ok()
+}
+
+pub(crate) async fn get_claims(store: &Store) -> Result<GetClaimsResponse> {
+    let mut claims = Vec::new();
+    let mut entries = store.keys().await?;
+    while let Some(key) = entries.next() {
+        if key.starts_with("CLAIMS_") {
+            add_claim(&mut claims, store.get(key).await?).await?;
+        }
+    }
+    Ok(GetClaimsResponse { claims: claims })
+}
+
+pub(crate) async fn get_links(store: &Store) -> Result<LinkDefinitionList> {
+    let mut links = Vec::new();
+    let mut entries = store.keys().await?;
+    while let Some(key) = entries.next() {
+        if key.starts_with("LINKDEF_") {
+            add_linkdef(&mut links, store.get(key).await?).await?;
+        }
+    }
+
+    Ok(LinkDefinitionList { links })
+}
+
+pub(crate) async fn put_link(store: &Store, ld: LinkDefinition) -> Result<()> {
+    let id = uuid::Uuid::new_v4().to_string();
+    let key = format!("LINKDEF_{}", id);
+    store
+        .put(key, serde_json::to_vec(&ld)?.into())
+        .await
+        .map(|_| ())
+}
+
+pub(crate) async fn delete_link(
+    store: &Store,
+    actor_id: &str,
+    contract_id: &str,
+    link_name: &str,
+) -> Result<()> {
+    if let Some(key) = find_link_key(store, actor_id, contract_id, link_name).await? {
+        store.delete(key).await.map(|_| ())
+    } else {
+        Err("No such link".into())
+    }
+}
+
+async fn add_linkdef(links: &mut Vec<LinkDefinition>, data: Option<Vec<u8>>) -> Result<()> {
+    if let Some(d) = data {
+        let ld: LinkDefinition = serde_json::from_slice(&d)?;
+        links.push(ld);
+    }
+
+    Ok(())
+}
+
+async fn find_link_key(
+    store: &Store,
+    actor_id: &str,
+    contract_id: &str,
+    link_name: &str,
+) -> Result<Option<String>> {
+    let mut entries = store.keys().await?;
+    while let Some(ref key) = entries.next() {
```

Comment:
This is O(n) now. I didn't realize link defs were keyed on a UUID and not a deterministic hash of the input. Are you concerned about performance for lattices with a lot of links, where every time we query links or delete a link, we construct a consumer that has to pull all values from the underlying stream one at a time?

Reply:
Agreed. The original problem was that I couldn't get both Rust and Elixir to agree upon the hash for the same link definition. I'll do some experimenting this morning and see.